lib/Scheduling/CPSATSchedulers.cpp

```diff
@@ -0,0 +1,154 @@
+  for (auto item : llvm::enumerate(tasks)) {
+    auto i = item.index();
+    auto *task = item.value();
+    IntVar startVar = cpModel.NewIntVar(Domain(0, horizon))
+                          .WithName((Twine("start_of_task_") + Twine(i)).str());
+    IntVar endVar = cpModel.NewIntVar(Domain(0, horizon))
+                        .WithName((Twine("end_of_task_") + Twine(i)).str());
+    taskStarts[task] = startVar;
+    taskEnds[task] = endVar;
+    auto resource = prob.getLinkedOperatorType(task);
+    unsigned duration = *prob.getLatency(*resource);
+    IntervalVar taskInterval =
+        cpModel.NewIntervalVar(startVar, duration, endVar)
+            .WithName((Twine("task_interval_") + Twine(i)).str());
+
+    if (prob.getLimit(*resource))
+      resourcesToTaskIntervals[resource.value()].emplace_back(taskInterval);
+  }
+
+  // Check for zero-distance cycles and otherwise establish operation ordering
+  // constraints.
+  for (Operation *task : tasks) {
+    for (auto dep : prob.getDependences(task)) {
+      Operation *src = dep.getSource();
+      Operation *dst = dep.getDestination();
+      if (src == dst)
+        return containingOp->emitError()
+               << "zero-distance dependence cycle detected";
+      cpModel.AddLessOrEqual(taskEnds[src], taskStarts[dst]);
+    }
+  }
+
+  // Establish "cumulative" constraints in order to constrain maximum
+  // concurrent usage of operators.
+  for (auto resourceToTaskIntervals : resourcesToTaskIntervals) {
+    Problem::OperatorType &resource = resourceToTaskIntervals.getFirst();
+    auto capacity = prob.getLimit(resource);
+    SmallVector<IntervalVar, 4> &taskIntervals =
+        resourceToTaskIntervals.getSecond();
+    // The semantics of cumulative constraints in or-tools are such that
+    // for any integer point, the sum of the demands across all
+    // intervals containing that point does not exceed the capacity of the
+    // resource. Thus tasks, in 1-1 correspondence with their intervals, are
+    // constrained to satisfy maximum resource requirements.
+    // See https://or.stackexchange.com/a/3363 for more details.
+    CumulativeConstraint cumu = cpModel.AddCumulative(capacity.value());
+    for (auto &item : llvm::enumerate(taskIntervals)) {
+      auto i = item.index();
+      auto taskInterval = item.value();
+      IntVar demandVar = cpModel.NewIntVar(Domain(1)).WithName(
+          (Twine("demand_") + Twine(i) + Twine("_") + Twine(resource.strref()))
+              .str());
+      // Conventional formulation for SharedOperatorsProblem;
+      // interval during which the resource is occupied has size 1.
+      IntervalVar start =
+          cpModel.NewFixedSizeIntervalVar(taskInterval.StartExpr(), 1);
+      cumu.AddDemand(start, demandVar);
+    }
+  }
+
+  IntVar makeSpan = cpModel.NewIntVar(Domain(0, horizon)).WithName("makespan");
+  cpModel.AddLessOrEqual(taskEnds[lastOp], makeSpan);
+  cpModel.Minimize(makeSpan);
```

Comment:
Can't you just `cpModel.Minimize(taskEnds[lastOp])`?

Reply:
yup